lib/puppet/pops/types/types.rb

```diff
@@ -666,12 +677,21 @@ def _assignable?(o, guard)
 # @api public
 #
 class PScalarType < PAnyType
+  KNOWN_SCALARS = [String, Integer, Float, TrueClass, FalseClass, Regexp, Time::Timestamp, Time::Timespan, SemanticPuppet::Version, SemanticPuppet::VersionRange]
+  KNOWN_NON_SCALARS = [Array, Hash, PAnyType, NilClass]
+
   def self.register_ptype(loader, ir)
     create_ptype(loader, ir, 'AnyType')
   end
 
   def instance?(o, guard = nil)
-    assignable?(TypeCalculator.infer(o), guard)
+    if KNOWN_SCALARS.any? { |scalar_class| o.is_a?(scalar_class) }
+      true
+    elsif KNOWN_NON_SCALARS.any? { |scalar_class| o.is_a?(scalar_class) }
+      false
```

Comment:
Could this be converted to a hash lookup instead of linear scan (albeit a short one), e.g

```ruby
KNOWN_SCALARS = { String => true, ...}
...
if KNOWN_SCALARS[scalar_class]
  true
...
```